File: felicia/examples/learn/service/grpc/python/grpc_client_node.py
import felicia_py as fel
import felicia_py.command_line_interface as cli
from felicia.core.protobuf.master_data_pb2 import ServiceInfo
from felicia.python.rpc.client import Client

import felicia.examples.learn.service.grpc.simple_service_pb2 as simple_service_pb2
import felicia.examples.learn.service.grpc.simple_service_pb2_grpc as simple_service_pb2_grpc


class SimpleClient(Client):
    @classmethod
    def new_stub(cls, channel):
        return simple_service_pb2_grpc.SimpleServiceStub(channel)

    def Add(self, request):
        return self.stub.Add(request)


class GrpcClientNode(fel.NodeLifecycle):
    def __init__(self, grpc_service_flag):
        super().__init__()
        self.grpc_service_flag = grpc_service_flag
        self.service = grpc_service_flag.service_flag.value
        self.simple_client = SimpleClient()
        self.client = fel.communication.ServiceClient(self.simple_client)

    def on_init(self):
        print("GrpcClientNode.on_init()")

    def on_did_create(self, node_info):
        print("GrpcClientNode.on_did_create()")
        self.node_info = node_info
        self.request_register()

    def on_error(self, status):
        print("GrpcClientNode.on_error()")
        fel.log(fel.ERROR, status.error_message())

    def on_request_register(self, status):
        print("GrpcClientNode.on_request_register()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

    def on_request_unregister(self, status):
        print("GrpcClientNode.on_request_unregister()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

    def on_connect(self, service_info_status):
        print("GrpcClientNode.on_connect()")
        if (service_info_status == ServiceInfo.REGISTERED):
            self.request_add()
        else:
            print("disconnected...")

    def request_register(self):
        self.client.request_register(
            self.node_info, self.service, self.on_connect, self.on_request_register)

    def request_unregister(self):
        self.client.request_unregister(
            self.node_info, self.service, self.on_request_unregister)

    def request_add(self):
        a = self.grpc_service_flag.a_flag.value
        b = self.grpc_service_flag.b_flag.value
        response = self.simple_client.Add(
            simple_service_pb2.AddRequest(a=a, b=b))
        print('{} + {} = {}'.format(a, b,
                                    cli.TextStyle.green(str(response.sum))))

File: felicia/examples/learn/service/grpc/cc/grpc_client_node.cc
#include "felicia/examples/learn/service/grpc/cc/grpc_client_node.h"

#include "third_party/chromium/base/strings/string_number_conversions.h"

#include "felicia/core/util/command_line_interface/text_style.h"

namespace felicia {

FEL_GRPC_CLIENT_METHOD_DEFINE(GrpcSimpleClient, Add)

GrpcClientNode::GrpcClientNode(const SimpleServiceFlag& simple_service_flag)
    : simple_service_flag_(simple_service_flag),
      service_(simple_service_flag.service_flag()->value()) {}

void GrpcClientNode::OnInit() {
  std::cout << "GrpcClientNode::OnInit()" << std::endl;
}

void GrpcClientNode::OnDidCreate(const NodeInfo& node_info) {
  std::cout << "GrpcClientNode::OnDidCreate()" << std::endl;
  node_info_ = node_info;
  RequestRegister();
}

void GrpcClientNode::OnError(const Status& s) {
  std::cout << "GrpcClientNode::OnError()" << std::endl;
  LOG(ERROR) << s;
}

void GrpcClientNode::OnRequestRegister(const Status& s) {
  std::cout << "GrpcClientNode::OnRequestRegister()" << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
}

void GrpcClientNode::OnRequestUnegister(const Status& s) {
  std::cout << "GrpcClientNode::OnRequestUnegister()" << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
}

void GrpcClientNode::OnConnect(ServiceInfo::Status s) {
  std::cout << "GrpcClientNode::OnConnect()" << std::endl;
  if (s == ServiceInfo::REGISTERED) {
    RequestAdd();
  } else {
    std::cout << "disconnected..." << std::endl;
  }
}

void GrpcClientNode::RequestRegister() {
  client_.RequestRegister(
      node_info_, service_,
      base::BindRepeating(&GrpcClientNode::OnConnect, base::Unretained(this)),
      base::BindOnce(&GrpcClientNode::OnRequestRegister,
                     base::Unretained(this)));
}

void GrpcClientNode::RequestUnregister() {
  client_.RequestUnregister(node_info_, service_,
                            base::BindOnce(&GrpcClientNode::OnRequestUnegister,
                                           base::Unretained(this)));
}

void GrpcClientNode::OnRequestAdd(const AddRequest* request,
                                  const AddResponse* response,
                                  const Status& s) {
  std::cout << "GrpcClientNode::OnRequestAdd()" << std::endl;
  if (s.ok()) {
    std::cout << request->a() << " + " << request->b() << " = "
              << TextStyle::Green(base::NumberToString(response->sum()))
              << std::endl;
  } else {
    LOG(ERROR) << s;
  }
}

void GrpcClientNode::RequestAdd() {
  AddRequest* request = new AddRequest();
  AddResponse* response = new AddResponse();
  request->set_a(simple_service_flag_.a_flag()->value());
  request->set_b(simple_service_flag_.b_flag()->value());
  client_->AddAsync(
      request, response,
      base::BindOnce(&GrpcClientNode::OnRequestAdd, base::Unretained(this),
                     base::Owned(request), base::Owned(response)));
}

}  // namespace felicia